ledger/src/blockstore.rs

```diff
@@ -1182,6 +1182,19 @@ impl Blockstore {
         let slot_meta = &mut slot_meta_entry.new_slot_meta.borrow_mut();
 
         if !is_trusted {
+            if shred.last_in_slot() && shred_index < slot_meta.received && !slot_meta.is_full() {
+                // We got a last shred < slot_meta.received, which signals there's an alternative,
+                // shorter version of the slot. Because also `!slot_meta.is_full()`, then this
+                // means, for the current version of the slot, we might never get all the
+                // shreds < the current last index, never replaythis slot, and make no
+                // progress (for instance if a leader sends an additional detached "last index"
+                // shred with a very high index, but none of the intermediate shreds). Ideally, we would
+                // just purge all shreds > the new last index slot, but because replay may have already
+                // replayed entries past the newly deteced "last" shred, then mark the slot as dead
+                // and wait for replay to dump and repair the correct version.
+                write_batch.put::<cf::DeadSlots>(slot, &true).unwrap();
```

Comment:
I added https://github.com/solana-labs/solana/pull/16071/commits/e8f58cae6871e79336717f1b35b18ec0f2f31e15 to prevent a potential race condition where a slot is marked dead after `get_slot_entries` checks if the slot is dead and before it actually fetches entries (which could be full). Race condition was detected and fixed in this test: https://github.com/solana-labs/solana/pull/16071/commits/5b49ba6605268d1dd851a5baa587fc8ba0e183d7